secure/key-manager/src/key_manager.rs

```diff
@@ -0,0 +1,161 @@
+    fn verify_key_not_created(&mut self, key_name: &str) -> Result<(), Error> {
+        match self.storage.get_public_key_for_name(key_name) {
+            Err(Error::KeyNotSet(_)) => Ok(()),
+            _ => Err(Error::KeyAlreadyExists(key_name.to_string())),
+        }
+    }
+
+    /// Rotates the consensus key by: (i) generating and storing a new consensus key in the
+    /// secure store; and (ii) generating and signing a transaction that updates the consensus key
+    /// for this validator node on the blockchain. If a failure occurs during the process (e.g.,
+    /// the consensus key does not exist, or a new transaction could not be generated), an error is
+    /// returned.
+    pub fn rotate_consensus_key(&mut self) -> Result<(), Error> {
+        // Ensure key has already been initialized before rotating
+        self.verify_key_correctly_initialized(CONSENSUS_KEY)?;
+
+        // Rotate the key in storage
+        let new_public_consensus_key = self.storage.rotate_key_pair(CONSENSUS_KEY)?;
+
+        // Update the on-chain validator config to reflect the new_public_consensus_key
+        let _ = self.update_on_chain_consensus_key(new_public_consensus_key);
+
+        Ok(())
+    }
+
+    /// Verifies that a named key has already been initialized correctly in the secure store.
+    /// Returns an error if the key doesn't exist, or if the value related to the key is not
+    /// correct.
+    fn verify_key_correctly_initialized(&mut self, key_name: &str) -> Result<(), Error> {
+        match self.storage.get_public_key_for_name(key_name) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// This method is called after KeyManager has failed but has now been restarted.
+    /// It verifies the internal state of secure storage and fixes any potential inconsistencies due
+    /// to the failure. For this, we: (i) verify that all keys in secure storage have already been
+    /// initialized. If not, we initialize these keys for the first time; and (ii) verify the
+    /// current consensus key registered on-chain reflects the state of secure storage. If not, we
+    /// broadcasts the current consensus key on-chain. On the next Validator reconfiguration event,
+    /// the consensus key will be registered and the node can begin participating in consensus once
+    /// again.
+    pub fn check_and_fix_keys_after_failure(&mut self) -> Result<(), Error> {
+        // TODO(joshlind): generate the correct policies for these keys (i.e., policies that respect
+        // the LSR and KeyManagement access rules)
+        let consensus_key_policy = Policy::public();
+        let validator_key_policy = Policy::public();
+
+        // Initialize both keys -- these calls  will fail if the keys have already been initialized!
+        let _ = self.initialize_key(CONSENSUS_KEY, &consensus_key_policy);
+        let _ = self.initialize_key(VALIDATOR_KEY, &validator_key_policy);
+
```

Comment:
I don't think you ever want to initialize again. If you get here and this actually succeeds, it means you had a catastrophic failure and likely should have humans intervene.